protected/humhub/services/MigrationService.php

```diff
@@ -0,0 +1,262 @@
+<?php
+
+/*
+ * @link      https://www.humhub.org/
+ * @copyright Copyright (c) 2023 HumHub GmbH & Co. KG
+ * @license   https://www.humhub.com/licences
+ */
+
+namespace humhub\services;
+
+use humhub\commands\MigrateController;
+use humhub\components\Event;
+use humhub\components\Module;
+use humhub\events\MigrationEvent;
+use Throwable;
+use Yii;
+use yii\base\ActionEvent;
+use yii\base\Component;
+use yii\base\Controller;
+use yii\base\InvalidConfigException;
+use yii\console\ExitCode;
+
+/**
+ * @since 1.16
+ */
+class MigrationService extends Component
+{
+    public const EVENT_AFTER_MIGRATION = 'afterMigration';
+    public const EVENT_BEFORE_MIGRATION = 'beforeMigration';
+    public const ACTION_ENABLE = 'enable';
+    public const ACTION_DISABLE = 'disable';
+    public const ACTION_UPDATE = 'update';
+    public const ACTION_MIGRATE = 'migrate';
+    protected const MIGRATION_UNINSTALL = 'uninstall';
+    protected const MIGRATION_UP = 'up';
+
+    protected Module $module;
+    protected ?string $path;
+
+    /**
+     * @param Module $module
+     * @param array $config = ['path' => string|null]
+     */
+    public function __construct(Module $module, array $config = [])
+    {
+        $this->module = $module;
+
+        $path = $config['path'] ?? null;
+        unset($config['path']);
+
+        if ($path === null) {
+            $path = $module->basePath . '/migrations';
+
+            if (!is_dir(realpath($path))) {
+                Yii::debug("Module has no migrations directory.", $this->module->id);
+                $path = null;
+            }
+        } elseif (false === realpath($path)) {
+            $message = sprintf(
+                "Migration path '%s' for module '%s' is invalid!",
+                $path,
+                $this->module->getName()
+            );
+            Yii::error($message, $this->module->id);
+            throw new \RuntimeException($message);
+        }
+
+        $this->path = $path;
+
+        parent::__construct($config);
+    }
+
+    public function getModule(): Module
+    {
+        return $this->module;
+    }
+
+    public function getPath(): ?string
+    {
+        return $this->path;
+    }
+
+    public function migrateUp(string $action): ?bool
```

Comment:
The current Module class has four methods that involve migrations:
1. `enable()`
2. `update()`
3. `migrate()`
4. `disable()`

To mirror those four "actions" I've added the `$action` parameter that is only used in the event (`MigrationEvent::$action`; so that registering callbacks may know what is the current action).

The "MigrationEvent::$migration" on the other hand shows the `up`, `down` and `uninstall`. And there's `new` to to check for new migrations.

Reply:
Hmm, I would actually like to remove this.

Can you give me UseCases for the `MigrateEvent`? And if there are any, that a differentiation between `enable`, `update`, `migrate` action  is necessary here?

If there is logic, could this not be simply implemented in the `Module::update()` method?

Reply:
I've tried to implement this along your suggestion:
https://github.com/humhub/humhub/pull/6550#issuecomment-1831587153

I have removed `MigrationEvent::$action`

Simple use case would be to clear a cache after a migration. Or to suspend the module upon migration error. Or to enable a feature in the settings upon migration success.

These events are basically a replacement for the current module's `migrate()` method.

I thought the long-term idea was to reduce as much code from the Module class as possible. Hence the migration code should be going into the `MigrationService` and later also the `enable()`, `update()` and `disable()` should be going into a `ModuleService`, which also adresses the issue that the updated code is not available after `update` if the class has already been loaded.

Having the functionality in the service, would no longer "allow" to add that code to methods in the module. Then module devs have two options:
- either register to the events called by those services
  - `MigrationService`: before and after migration
  - `ModuleService`: before and after install/enable/disable/uninstall plus after update (so the class is not loaded before the file is replaced)
- alternatively, they may overwrite the `getMigrationService()` and create their own implementation of the service.

Hence, the events are meant to avoid the need to create a custom service class, but allow to just register to the events.

Reply:
@martin-rueegg Thank you.

Maybe we misunderstood a bit.

I think we should have implementable (enable/disable/update) methods in the module base class for the module developer. So that the workaround via events etc. is not necessary.

We should only move the core code that is necessary for migrating/enabling/updating into services.

Reply:
I see.

I thought the main goal was to reduce the code that needs to be loaded on most requests (as the module classes are usually loaded). Specifically code that is hardly ever used (as enable/migrate/update/disable) should be outsourced, so it is only loaded if required.

If we keep the enable/update/disable methods in the class, than we do not gain much, as the current methods in the base class were very light-weight and the heavy-lifting for a module's disable procedure, for example, is in the dev's concrete implementation. Getting this away would serve us much more than the few lines we've now saved with the `MigrationService`.

Reply:
My goal was more improved logical code separation for better maintainability.
I really like it, that the Migration code is now in a separate service.

I currently tend to leave enable/disable/update methods in the module class. I assume that all classes are already loaded and in the PHP cache on application servers.